src/Stratis.Bitcoin.Features.Consensus/PowConsensusValidator.cs

```diff
@@ -46,32 +58,216 @@ public PowConsensusValidator(Network network, ICheckpoints checkpoints, IDateTim
             this.Checkpoints = checkpoints;
         }
 
-        public virtual void CheckBlockHeader(ContextInformation context)
+        /// <summary>
+        /// Calculates total signature operation cost of a transaction.
+        /// </summary>
+        /// <param name="transaction">Transaction for which we are computing the cost.</param>
+        /// <param name="inputs">Map of previous transactions that have outputs we're spending.</param>
+        /// <param name="flags">Script verification flags.</param>
+        /// <returns>Signature operation cost for all transaction's inputs.</returns>
+        public long GetTransactionSignatureOperationCost(Transaction transaction, UnspentOutputSet inputs, DeploymentFlags flags)
         {
-            if (context.CheckPow && !context.BlockValidationContext.Block.Header.CheckProofOfWork())
-                ConsensusErrors.HighHash.Throw();
+            long signatureOperationCost = this.GetLegacySignatureOperationsCount(transaction) * this.ConsensusOptions.WitnessScaleFactor;
 
-            context.NextWorkRequired = context.BlockValidationContext.ChainedBlock.GetWorkRequired(context.Consensus);
+            if (transaction.IsCoinBase)
+                return signatureOperationCost;
+
+            if (flags.ScriptFlags.HasFlag(ScriptVerify.P2SH))
+            {
+                signatureOperationCost += this.GetP2SHSignatureOperationsCount(transaction, inputs) * this.ConsensusOptions.WitnessScaleFactor;
+            }
+
+            for (int i = 0; i < transaction.Inputs.Count; i++)
+            {
+                TxOut prevout = inputs.GetOutputFor(transaction.Inputs[i]);
+                signatureOperationCost += this.CountWitnessSignatureOperation(transaction.Inputs[i].ScriptSig, prevout.ScriptPubKey, transaction.Inputs[i].WitScript, flags);
+            }
+
+            return signatureOperationCost;
         }
 
+        /// <summary>
+        /// Checks that transaction's inputs are valid.
+        /// </summary>
+        /// <param name="transaction">Transaction to check.</param>
+        /// <param name="inputs">Map of previous transactions that have outputs we're spending.</param>
+        /// <param name="spendHeight">Height at which coins are considered spendable.</param>
```

Comment:
i think this is incorrect
this seems to be the height at which we are spending
not height at which the coins are spendable
CheckMaturity call will tell us if those are spendable